pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1843,18 +1812,29 @@ private boolean processPossibleToDLQ(MessageIdImpl messageId) {
                 }
             }
             if (deadLetterProducer != null) {
-                try {
-                    for (MessageImpl<T> message : deadLetterMessages) {
-                        deadLetterProducer.newMessage()
-                                .value(message.getValue())
-                                .properties(message.getProperties())
-                                .send();
-                    }
-                    acknowledge(messageId);
-                    return true;
-                } catch (Exception e) {
-                    log.error("Send to dead letter topic exception with topic: {}, messageId: {}", deadLetterProducer.getTopic(), messageId, e);
+                for (MessageImpl<T> message : deadLetterMessages) {
+                    deadLetterProducer.newMessage()
+                            .value(message.getValue())
+                            .properties(message.getProperties())
+                            .sendAsync()
+                            .whenComplete((msgId, ex) -> {
+                                if (ex != null) {
+                                    log.error("Send to dead letter topic exception with topic: {}, messageId: {}",
+                                            deadLetterProducer.getTopic(), messageId, ex);
+                                    return;
+                                }
+
+                                log.debug("Redeliver dead letter message topic: {}, messageId: {} to dead letter topic with topic: {}, messageId: {}",
+                                        message.getTopicName(), messageId, deadLetterProducer.getTopic(), msgId);
+                                try {
+                                    acknowledge(messageId);
+                                } catch (Exception e) {
+                                    log.error("Acknowledge dead letter message exception with topic: {}, messageId: {}",
+                                            message.getTopicName(), messageId);
+                                }
+                            });
                 }
+                return true;
```

Comment:
You can't return true here because this will always return true even if the message write to the dead letter topic fails. This is used to filter out the messages that should not be delivered.

I think maybe you should change the method `processPossibleToDLQ` to `processPossibleToDLQAsync`.

Reply:
Right, can't simply return true here, but make the method `processPossibleToDLQ` asynchronous will bring many other changes to fit async.
https://github.com/apache/pulsar/blob/7a9ec066e1cc3ca918240776c4c9f805b0a92e96/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L1787-L1794
https://github.com/apache/pulsar/blob/7a9ec066e1cc3ca918240776c4c9f805b0a92e96/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L701-L703

How about re-put message into the `UnAckedMessageTracker` when sent failed? So these messages will be retry in later tick.

Reply:
It's not a safe operation since the client might crash before the next retry.